File: services/products/src/test/java/br/com/deltaglobalbank/products/features/lending/listLendingProduct/ListLendingProductsTests.java
package br.com.deltaglobalbank.products.features.lending.listLendingProduct;

import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.products.TestcontainersConfiguration;
import br.com.deltaglobalbank.products.domain.product.AgreementName;
import br.com.deltaglobalbank.products.domain.product.DisplayName;
import br.com.deltaglobalbank.products.domain.product.Product;
import br.com.deltaglobalbank.products.domain.product.ProductType;
import br.com.deltaglobalbank.products.infrastructure.persistence.mappers.ProductMapper;
import br.com.deltaglobalbank.products.infrastructure.persistence.repositories.JpaProductRepository;
import br.com.deltaglobalbank.sharedauth.AuthenticatedPrincipal;
import br.com.deltaglobalbank.sharedauth.JwtAuthenticationToken;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;

@SpringBootTest
@ActiveProfiles("test")
@AutoConfigureMockMvc
@Import(TestcontainersConfiguration.class)
@Transactional
class ListLendingProductsTests {

    @Autowired
    MockMvc mockMvc;
    @Autowired
    JpaProductRepository jpaProductRepository;

    private final UUID tenantId = UUID.randomUUID();

    private JwtAuthenticationToken principal(String... roles) {
        return principalForTenant(tenantId, roles);
    }

    private JwtAuthenticationToken principalForTenant(UUID tenant, String... roles) {
        return new JwtAuthenticationToken(
            new AuthenticatedPrincipal(
                UUID.randomUUID(), tenant, "user",
                List.of(roles), List.of(),
                false, UUID.randomUUID()
            )
        );
    }

    private UUID seedProduct(UUID tenant, String agreementName, boolean active) {
        UUID id = UUID.randomUUID();
        Instant now = Instant.now();
        Product product = new Product(
            id, tenant, ProductType.LENDING,
            new AgreementName(agreementName), new DisplayName("Produto"),
            new BigDecimal("1.5"), new BigDecimal("3.0"),
            12, 60,
            new BigDecimal("1000.00"), new BigDecimal("50000.00"),
            new BigDecimal("2.0"), active,
            now, now, null, null
        );
        jpaProductRepository.save(ProductMapper.toEntity(product));
        return id;
    }

    private UUID seedProduct() {
        return seedProduct(tenantId, "INSS", true);
    }

    private UUID seedProduct(String agreementName) {
        return seedProduct(tenantId, agreementName, true);
    }

    private UUID seedProduct(String agreementName, boolean active) {
        return seedProduct(tenantId, agreementName, active);
    }

    private UUID seedProduct(UUID tenant, String agreementName) {
        return seedProduct(tenant, agreementName, true);
    }

    @Test
    void returnsAllProductsWithPaginationWrapper() throws Exception {
        for (int i = 0; i < 3; i++) {
            seedProduct("INSS-" + i);
        }
        mockMvc.perform(get("/products/lending")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(3))
            .andExpect(jsonPath("$.page").value(0))
            .andExpect(jsonPath("$.totalElements").value(3));
    }

    @Test
    void paginatesWithPageAndSizeParams() throws Exception {
        for (int i = 0; i < 3; i++) {
            seedProduct("INSS-" + i);
        }
        mockMvc.perform(get("/products/lending?page=0&size=2")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(2))
            .andExpect(jsonPath("$.page").value(0))
            .andExpect(jsonPath("$.size").value(2))
            .andExpect(jsonPath("$.totalElements").value(3))
            .andExpect(jsonPath("$.totalPages").value(2));
    }

    @Test
    void ordersByCreatedAtDescending() throws Exception {
        UUID firstId = seedProduct("AAA");
        Thread.sleep(10);
        UUID lastId = seedProduct("BBB");
        mockMvc.perform(get("/products/lending")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items[0].id").value(lastId.toString()));
    }

    @Test
    void coercesSizeAboveMaxTo100() throws Exception {
        seedProduct();
        mockMvc.perform(get("/products/lending?size=9999")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.size").value(100));
    }

    @Test
    void usesDefaultSize20WhenOmitted() throws Exception {
        seedProduct();
        mockMvc.perform(get("/products/lending")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.size").value(20));
    }

    @Test
    void returns401WithoutToken() throws Exception {
        mockMvc.perform(get("/products/lending"))
            .andExpect(status().isUnauthorized());
    }

    @Test
    void returnsEmptyListForTenantWithoutProducts() throws Exception {
        mockMvc.perform(get("/products/lending")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(0))
            .andExpect(jsonPath("$.totalElements").value(0));
    }

    @Test
    void filtersByAgreementName() throws Exception {
        seedProduct("INSS");
        seedProduct("SIAPE");
        mockMvc.perform(get("/products/lending?agreementName=INSS")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(1))
            .andExpect(jsonPath("$.items[0].agreementName").value("INSS"));
    }

    @Test
    void filtersByAgreementNameCaseInsensitively() throws Exception {
        seedProduct("INSS");
        mockMvc.perform(get("/products/lending?agreementName=inss")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(1));
    }

    @Test
    void filtersByAgreementNameSubstring() throws Exception {
        seedProduct("INSS");
        mockMvc.perform(get("/products/lending?agreementName=NS")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(1))
            .andExpect(jsonPath("$.items[0].agreementName").value("INSS"));
    }

    @Test
    void filtersActiveTrue() throws Exception {
        seedProduct("ATIVO", true);
        seedProduct("INATIVO", false);
        mockMvc.perform(get("/products/lending?active=true")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(1))
            .andExpect(jsonPath("$.items[0].active").value(true));
    }

    @Test
    void filtersActiveFalse() throws Exception {
        seedProduct("ATIVO", true);
        seedProduct("INATIVO", false);
        mockMvc.perform(get("/products/lending?active=false")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(1))
            .andExpect(jsonPath("$.items[0].active").value(false));
    }

    @Test
    void returnsBothActiveAndInactiveWhenActiveFilterIsAbsent() throws Exception {
        seedProduct("ATIVO", true);
        seedProduct("INATIVO", false);
        mockMvc.perform(get("/products/lending")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(2));
    }

    @Test
    void combinesAgreementNameAndActiveFilters() throws Exception {
        seedProduct("INSS", true);
        seedProduct("INSS", false);
        seedProduct("SIAPE", true);
        mockMvc.perform(get("/products/lending?agreementName=INSS&active=true")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.items.length()").value(1))
            .andExpect(jsonPath("$.items[0].agreementName").value("INSS"))
            .andExpect(jsonPath("$.items[0].active").value(true));
    }

    @Test
    void listsOnlyOwnTenantProducts() throws Exception {
        UUID tenantA = tenantId;
        UUID tenantB = UUID.randomUUID();
        seedProduct(tenantA, "A1");
        seedProduct(tenantA, "A2");
        seedProduct(tenantB, "B1");
        mockMvc.perform(get("/products/lending")
                .with(authentication(principal("products.lending.read"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.totalElements").value(2));
    }

    @Test
    void platformAdminListsProductsOfASpecificTenant() throws Exception {
        UUID tenantA = UUID.randomUUID();
        seedProduct(tenantA, "A1");
        seedProduct(tenantA, "A2");
        seedProduct(UUID.randomUUID(), "OTHER");
        mockMvc.perform(get("/products/tenants/" + tenantA + "/lending")
                .with(authentication(principal("platform.admin"))))
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.totalElements").value(2));
    }
}

File: services/products/src/test/kotlin/br/com/deltaglobalbank/products/features/lending/listLendingProduct/ListLendingProductsTests.kt
package br.com.deltaglobalbank.products.features.lending.listLendingProduct

import br.com.deltaglobalbank.products.TestcontainersConfiguration
import br.com.deltaglobalbank.products.domain.product.*
import br.com.deltaglobalbank.products.infrastructure.persistence.mappers.toEntity
import br.com.deltaglobalbank.products.infrastructure.persistence.repositories.JpaProductRepository
import br.com.deltaglobalbank.sharedauth.AuthenticatedPrincipal
import br.com.deltaglobalbank.sharedauth.JwtAuthenticationToken
import jakarta.transaction.Transactional
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.context.annotation.Import
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.authentication
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import java.math.BigDecimal
import java.util.UUID
import kotlin.test.Test

@SpringBootTest
@ActiveProfiles("test")
@AutoConfigureMockMvc
@Import(TestcontainersConfiguration::class)
@Transactional
class ListLendingProductsTests {

    @Autowired lateinit var mockMvc: MockMvc
    @Autowired lateinit var jpaProductRepository: JpaProductRepository

    private val tenantId = UUID.randomUUID()

    private fun principal(vararg roles: String) = principalForTenant(tenantId, *roles)

    private fun principalForTenant(tenant: UUID, vararg roles: String) = JwtAuthenticationToken(
        AuthenticatedPrincipal(
            subject = UUID.randomUUID(), tenantId = tenant, principalType = "user",
            roles = roles.toList(), modules = emptyList(),
            mustChangePassword = false, jti = UUID.randomUUID(),
        )
    )

    private fun seedProduct(
        tenant: UUID = tenantId,
        agreementName: String = "INSS",
        active: Boolean = true,
    ): UUID {
        val id = UUID.randomUUID()
        val now = java.time.Instant.now()
        val product = Product(
            id = id, tenantId = tenant, type = ProductType.LENDING,
            agreementName = AgreementName(agreementName), displayName = DisplayName("Produto"),
            minMonthlyRate = BigDecimal("1.5"), maxMonthlyRate = BigDecimal("3.0"),
            minMonths = 12, maxMonths = 60,
            minAmount = BigDecimal("1000.00"), maxAmount = BigDecimal("50000.00"),
            commissionRate = BigDecimal("2.0"), active = active,
            createdAt = now, updatedAt = now, createdBy = null, updatedBy = null,
        )
        jpaProductRepository.save(product.toEntity())
        return id
    }

    @Test
    fun `returns all products with pagination wrapper`() {
        repeat(3) { seedProduct(agreementName = "INSS-$it") }
        mockMvc.get("/products/lending") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(3) }
            jsonPath("$.page") { value(0) }
            jsonPath("$.totalElements") { value(3) }
        }
    }

    @Test
    fun `paginates with page and size params`() {
        repeat(3) { seedProduct(agreementName = "INSS-$it") }
        mockMvc.get("/products/lending?page=0&size=2") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(2) }
            jsonPath("$.page") { value(0) }
            jsonPath("$.size") { value(2) }
            jsonPath("$.totalElements") { value(3) }
            jsonPath("$.totalPages") { value(2) }
        }
    }

    @Test
    fun `orders by createdAt descending`() {
        val firstId = seedProduct(agreementName = "AAA")
        Thread.sleep(10)
        val lastId = seedProduct(agreementName = "BBB")
        mockMvc.get("/products/lending") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items[0].id") { value(lastId.toString()) }
        }
    }

    @Test
    fun `coerces size above max to 100`() {
        seedProduct()
        mockMvc.get("/products/lending?size=9999") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.size") { value(100) }
        }
    }

    @Test
    fun `uses default size 20 when omitted`() {
        seedProduct()
        mockMvc.get("/products/lending") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.size") { value(20) }
        }
    }

    @Test
    fun `returns 401 without token`() {
        mockMvc.get("/products/lending")
            .andExpect { status { isUnauthorized() } }
    }

    @Test
    fun `returns empty list for tenant without products`() {
        mockMvc.get("/products/lending") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(0) }
            jsonPath("$.totalElements") { value(0) }
        }
    }

    @Test
    fun `filters by agreement name`() {
        seedProduct(agreementName = "INSS")
        seedProduct(agreementName = "SIAPE")
        mockMvc.get("/products/lending?agreementName=INSS") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(1) }
            jsonPath("$.items[0].agreementName") { value("INSS") }
        }
    }

    @Test
    fun `filters by agreement name case-insensitively`() {
        seedProduct(agreementName = "INSS")
        mockMvc.get("/products/lending?agreementName=inss") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(1) }
        }
    }

    @Test
    fun `filters by agreement name substring`() {
        seedProduct(agreementName = "INSS")
        mockMvc.get("/products/lending?agreementName=NS") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(1) }
            jsonPath("$.items[0].agreementName") { value("INSS") }
        }
    }

    @Test
    fun `filters active true`() {
        seedProduct(agreementName = "ATIVO", active = true)
        seedProduct(agreementName = "INATIVO", active = false)
        mockMvc.get("/products/lending?active=true") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(1) }
            jsonPath("$.items[0].active") { value(true) }
        }
    }

    @Test
    fun `filters active false`() {
        seedProduct(agreementName = "ATIVO", active = true)
        seedProduct(agreementName = "INATIVO", active = false)
        mockMvc.get("/products/lending?active=false") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(1) }
            jsonPath("$.items[0].active") { value(false) }
        }
    }

    @Test
    fun `returns both active and inactive when active filter is absent`() {
        seedProduct(agreementName = "ATIVO", active = true)
        seedProduct(agreementName = "INATIVO", active = false)
        mockMvc.get("/products/lending") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(2) }
        }
    }

    @Test
    fun `combines agreement name and active filters`() {
        seedProduct(agreementName = "INSS", active = true)
        seedProduct(agreementName = "INSS", active = false)
        seedProduct(agreementName = "SIAPE", active = true)
        mockMvc.get("/products/lending?agreementName=INSS&active=true") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.items.length()") { value(1) }
            jsonPath("$.items[0].agreementName") { value("INSS") }
            jsonPath("$.items[0].active") { value(true) }
        }
    }

    @Test
    fun `lists only own tenant products`() {
        val tenantA = tenantId
        val tenantB = UUID.randomUUID()
        seedProduct(tenant = tenantA, agreementName = "A1")
        seedProduct(tenant = tenantA, agreementName = "A2")
        seedProduct(tenant = tenantB, agreementName = "B1")
        mockMvc.get("/products/lending") {
            with(authentication(principal("products.lending.read")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.totalElements") { value(2) }
        }
    }

    @Test
    fun `platform admin lists products of a specific tenant`() {
        val tenantA = UUID.randomUUID()
        seedProduct(tenant = tenantA, agreementName = "A1")
        seedProduct(tenant = tenantA, agreementName = "A2")
        seedProduct(tenant = UUID.randomUUID(), agreementName = "OTHER")
        mockMvc.get("/products/tenants/$tenantA/lending") {
            with(authentication(principal("platform.admin")))
        }.andExpect {
            status { isOk() }
            jsonPath("$.totalElements") { value(2) }
        }
    }
}